storage2/src/state.rs

```diff
@@ -29,14 +34,22 @@ impl State {
     pub fn begin_transaction(&mut self) -> StateTransaction {
         StateTransaction::new(self)
     }
+
+    // Apply the unwritten changes of a transaction to this state fork.
+    pub fn apply_transaction(&mut self, transaction: StateTransaction) {
+        for (key, value) in transaction.unwritten_changes.iter() {
+            self.unwritten_changes.insert(key.clone(), value.clone());
+        }
+    }
 }
 
 #[async_trait]
 impl StateRead for State {
     fn get_raw(&self, key: String) -> Option<Vec<u8>> {
         // If the key is available in the unwritten_changes cache, return it.
+        // A `None` value represents that the key has been deleted.
         if let Some(value) = self.unwritten_changes.get(&key) {
-            return Some(value.clone());
+            return value.clone();
```

Comment:
I got confused reading this, because I thought that
```
// A `None` value ...
```
was referring to the return value of `.get(&key)`, not to the *inner* value.  It might be clearer to say something like

> `.get()` returns an `Option<Option<Vec<u8>>`; the outer `Option` signals whether the key is in the `unwritten_changes` cache, and the inner `Option` signals whether the value is present or deleted

Reply:
You could also write:

```rust
fn get_raw(&self, key: String) -> Option<Vec<u8>> {
    self.unwritten_changes.get(&key).or_else(|| self.snapshot.get_raw(key))
}
```

This makes it more evident to me what's going on here.